### PR TITLE
Integrate with batch transactions

### LINKED ISSUE #23: This cctp relayer should handle batch messages

### PR DESCRIPTION
**Pull request by boojamya**
Batch tx example: https://etherscan.io/tx/0x55dcde141614c0baf34827d1dcdced183b2ea166cec0b849a5c8eb029ec0bb92

Prior to this PR, the relayer couldn't handle batch transactions. 
As you can see here:
https://github.com/strangelove-ventures/noble-cctp-relayer/blob/bd7b450e96529bfa75f8e2274591424fa94d11c3/cmd/process.go#L92-L94

We were storing and loading messages by a lookup key created by the `SourceTxHash` and the `Type`, however batch transactions could have the same `SourceTxHash` and the `Type` but different message `MsgSendBytes`.

So as a transactions with multiple messages would come in, they would overwrite each other in the `processingQueue`

### CODE REVIEW

File: cmd/noble/broadcast.go [hunk 1]
```diff
@@ -227,7 +227,7 @@ func GetNobleAccountNumberSequence(urlBase string, address string) (int64, int64
 	var resp types.AccountResp
 	err = json.Unmarshal(body, &resp)
 	if err != nil {
-		return 0, 0, errors.New("unable to parse account number, sequence")
+		return 0, 0, fmt.Errorf("unable to parse account number, sequence. Raw HHTP Get response: %s", string(body))
```
Comment (jtieri) [inline cmd/noble/broadcast.go]: nit: this comma feels oddly out of place

File: cmd/process.go [hunk 2]
```diff
@@ -85,84 +96,123 @@ func Start(cmd *cobra.Command, args []string) {
...
 		if !ok {
-			State.Store(LookupKey(dequeuedMsg.SourceTxHash, dequeuedMsg.Type), dequeuedMsg)
-			msg, _ = State.Load(LookupKey(dequeuedMsg.SourceTxHash, dequeuedMsg.Type))
-			msg.Status = types.Created
+			batchPosition = 0
+			State.Store(dequeuedMsg.SourceTxHash, []*types.MessageState{dequeuedMsg})
+			msg, _ = State.Load(dequeuedMsg.SourceTxHash)
+			p.Mu.Lock()
+			msg[batchPosition].Status = types.Created
+			p.Mu.Unlock()
```
Comment (jtieri) [inline cmd/process.go]: why does the `MessageState` need to be stored, then loaded before mutating the msg at index `0` to set `Status = types.Created`?

couldn't you set `Status = types.Created` before calling `Store`?

File: cmd/process.go [hunk 3]
```diff
@@ -85,84 +96,123 @@ func Start(cmd *cobra.Command, args []string) {
...
+			// if we have seen this transaction hash, check if we have seen this message
+			// if not, add it to the messageState slice tied to the tx hash
+			found := false
+			for pos, s := range msg {
+				if s == dequeuedMsg {
+					batchPosition = pos
+					found = true
+					break
+				}
+			}
```
Comment (jtieri) [inline cmd/process.go]: it looks like this if statement is checking the equality of two pointer types which may not have the intended behavior here. see: https://stackoverflow.com/questions/34276205/comparing-pointers-in-go

`dequeuedMsg` is read from the `processingQueue` channel, which i'm guessing is reading data from over the network, and `s` is loaded from state so even if the two underlying structs have the same data fields the pointers are pointing to different locations in memory.

to properly check the equality of two different instances of `MessageState` you will want to make use of an `Equals` method which explicitly checks the equality of each field in the structs.

File: cmd/process.go [hunk 4]
```diff
@@ -85,84 +96,123 @@ func Start(cmd *cobra.Command, args []string) {
...
-		// if this is the first time seeing this message, add it to the State
-		msg, ok := State.Load(LookupKey(dequeuedMsg.SourceTxHash, dequeuedMsg.Type))
+
+		var batchPosition int
+		// if this is the first time seeing this tx hash, add the message to State
+		msg, ok := State.Load(dequeuedMsg.SourceTxHash)
 		if !ok {
-			State.Store(LookupKey(dequeuedMsg.SourceTxHash, dequeuedMsg.Type), dequeuedMsg)
-			msg, _ = State.Load(LookupKey(dequeuedMsg.SourceTxHash, dequeuedMsg.Type))
-			msg.Status = types.Created
```
Comment (jtieri) [inline cmd/process.go:98]: should `msg.Status = types.Created` still be applied in the case that the msg was not found for the specified tx hash?

File: cmd/process.go [hunk 5]
```diff
@@ -85,84 +96,123 @@ func Start(cmd *cobra.Command, args []string) {
...
 			if response != nil {
-				if msg.Status == types.Created && response.Status == "pending_confirmations" {
-					msg.Status = types.Pending
-					msg.Updated = time.Now()
+				if msg[batchPosition].Status == types.Created && response.Status == "pending_confirmations" {
+					p.Mu.Lock()
+					msg[batchPosition].Status = types.Pending
+					msg[batchPosition].Updated = time.Now()
+					p.Mu.Unlock()
 					time.Sleep(10 * time.Second)
```
Comment (jtieri) [inline cmd/process.go]: out of curiosity why do we need to wait 10 seconds before sending the msg over the `processingQueue` channel?

File: cmd/process_test.go [hunk 6]
```diff
@@ -53,11 +56,13 @@ func TestProcessNewLog(t *testing.T) {
...
-	time.Sleep(2 * time.Second)
+	time.Sleep(5 * time.Second)
 
-	actualState, _ := cmd.State.Load(cmd.LookupKey(expectedState.SourceTxHash, expectedState.Type))
+	actualState, _ := cmd.State.Load(expectedState.SourceTxHash)
 
-	require.Equal(t, types.Created, actualState.Status)
+	p.Mu.RLock()
+	require.Equal(t, types.Created, actualState[0].Status)
+	p.Mu.RUnlock()
```
Comment (jtieri) [inline cmd/process_test.go]: why is the lock necessary in this test case?

File: cmd/process_test.go [hunk 7]
```diff
@@ -83,10 +90,11 @@ func TestProcessCreatedLog(t *testing.T) {
...
 	time.Sleep(2 * time.Second)
 
-	actualState, ok := cmd.State.Load(cmd.LookupKey(expectedState.SourceTxHash, expectedState.Type))
+	actualState, ok := cmd.State.Load(expectedState.SourceTxHash)
 	require.True(t, ok)
-	require.Equal(t, types.Complete, actualState.Status)
-
+	p.Mu.RLock()
+	require.Equal(t, types.Complete, actualState[0].Status)
+	p.Mu.RUnlock()
```
Comment (jtieri) [inline cmd/process_test.go]: see previous question about locks being used in these test cases

File: types/state_test.go [hunk 8]
```diff
@@ -1,28 +1,45 @@
...
-	stateMap.Store("123", &msg)
 
-	lMsg, _ := stateMap.Load("123")
-	fmt.Println(lMsg)
+	txHash := "123456789"
+	msg := MessageState{
+		SourceTxHash: txHash,
+		IrisLookupId: "123",
+		Status:       Filtered,
+		MsgSentBytes: []byte("i like turtles"),
```
Comment (jtieri) [inline types/state_test.go:17]: 😂 🐢

**Review by jtieri — COMMENTED** [review]
had a couple nits, a few questions and i think just one real issue i pointed out regarding the pointer equality stuff!

**Review by agouin — APPROVED** [review]

**Pull request merged**